pennylane/templates/decorator.py

```diff
@@ -54,10 +54,18 @@ def circuit():
     Returns:
         callable: The wrapper function
     """
+    # pylint: disable=import-outside-toplevel
 
     @wraps(func)
     def wrapper(*args, **kwargs):
-        with OperationRecorder() as rec:
+        import pennylane as qml
+
+        recorder_class = OperationRecorder
+
+        if qml.tape_mode_active():
+            recorder_class = qml.tape.TapeOperationRecorder
+
+        with recorder_class() as rec:
```

Comment:
Just to understand, why is it not

```python
     if qml.tape_mode_active():
            import pennylane as qml
            recorder_class = qml.tape.TapeOperationRecorder
     else: 
        recorder_class = OperationRecorder
```

Reply:
If tape mode is active to reimport, else you use the old one?

Reply:
If PennyLane is already imported in the Python session, `import pennylane as qml` won't actually do anything (e.g., the `__init__.py` won't be re-executed).

The `import pennylane as qml` is just to make the `qml` namespace available for this session, so that we can access the `tape_mode_active` and `TapeOperationRecorder` functions/classes.

I couldn't include this import at the top of the file due to circular imports, so that's the only reason it is inside the decorator :(

Reply:
Makes sense. I never understood circular imports, maybe we should solve the issue once and for all...